### PR TITLE
[zeus] script phase: "complete all checkpoints" upgrade script for slashing

### PR DESCRIPTION
**Pull request by jbrower95**
# Summary
- This go binary (intended to be run during the `script` phase of our [slashing upgrade](https://github.com/Layr-Labs/eigenlayer-contracts/pull/916)) attempts to;
   - Identify all the EigenPods on the network
   - Identify which of these EigenPods have out-standing checkpoints
   - Generate proofs + complete those checkpoints.
- This script phase will only be run if the ZEUS_ENV indicates that the deployment of EigenLayer supports EigenPods.

# Methodology
- Compute `V`, the set of all validators who withdrawal addresses are pointed at the execution chain.
- Check the set of `V`'s withdrawal addresses to see which ones may be EigenPods.
    - cast to EigenPod, check `podOwner()` that it claims, and then check the canonical EigenPodManager to see if `ownerToPod()` matches. 
- Check each EigenPod for currentCheckpointTimestamp > 0
- Complete the checkpoint (follows the same codepath as our eigenpod-proofs-generation CLI, which has been in prod for 

# Arguments
**$SENDER_PK**: A private key for an ethereum wallet containing sufficient funds to complete all of the checkpoints. This could be several ETH depending on how big the oustanding checkpoints are.
**$RPC_URL**: An ETH RPC to the network the deploy is running on.
**$BEACON_URL**: An ETH2 Beacon RPC to the network the deploy is running on.

# Example transaction
- [0x9eb4b54ef4b2a961feb2b6f359df0640ff9d085cd23a21a33aaa116c1c406ef2](https://holesky.etherscan.io/tx/0x9eb4b54ef4b2a961feb2b6f359df0640ff9d085cd23a21a33aaa116c1c406ef2)

# Repro 
`SENDER_PK=$EIGENPOD_PK RPC_URL=$NODE_ETH BEACON_URL=$NODE_ETH_HOLESKY_2 zeus run --env preprod -c "go run script.go"`

### CODE REVIEW

File: script/releases/v1.0.0-slashing/cleanup/script.go [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+	panicOnError("failed to reach beacon node", err)
+
+	panicOnError("failed to reach ethereum clients", err)
+
+	mc, err := multicall.NewMulticallClient(ctx, eth, &multicall.TMulticallClientOptions{
+		MaxBatchSizeBytes: 8192,
+	})
+	panicOnError("error initializing mc", err)
+
+	podManagerAddress := os.Getenv("ZEUS_DEPLOYED_EigenPodManager_Proxy")
```
Comment (gpsanant) [inline script/releases/v1.0.0-slashing/cleanup/script.go:105]: misc: a zeus helper lib might be a nice thing in the future

File: script/releases/v1.0.0-slashing/cleanup/script.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+	// fetch latest beacon state.
+	_validators := (func() *map[phase0.ValidatorIndex]*v1.Validator {
+		if provider, isProvider := beaconClient.(eth2client.ValidatorsProvider); isProvider {
```
Comment (gpsanant) [inline script/releases/v1.0.0-slashing/cleanup/script.go:109]: we should panic if not a provider

File: script/releases/v1.0.0-slashing/cleanup/script.go [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+			})
+			panicOnError("failed to load validator set", err)
+			return &validators.Data
+		}
+		return nil
+	})()
+	if _validators == nil {
+		panic("failed to load validators")
+	}
+	validators := *_validators
```
Comment (gpsanant) [inline script/releases/v1.0.0-slashing/cleanup/script.go:124]: can we log the number of validators, just to sanity check?

File: script/releases/v1.0.0-slashing/cleanup/script.go [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+	coreBeaconClient, _, err := proofgen.NewBeaconClient(args.BeaconNode, true /* verbose */)
+	panicOnError("failed to instantiate beaconClient", err)
+
+	for i := 0; i < len(results); i++ {
+		fmt.Printf("Completing [%d/%d]...", i+1, len(results))
+		fmt.Printf("NOTE: this is expensive, and may take several minutes.")
+		completeCheckpointForEigenpod(ctx, results[i].Address, eth, chainId, coreBeaconClient, args.Sender)
+	}
+
+	return nil
```
Comment (gpsanant) [inline script/releases/v1.0.0-slashing/cleanup/script.go:184]: can we sanity check that all checkpoint timestamps are zero

File: script/releases/v1.0.0-slashing/cleanup/script.go [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+	})
+
+	interestingWithdrawalAddresses := lo.Keys(lo.Reduce(executionLayerWithdrawalCredentialValidators, func(accum map[string]int, next ValidatorWithIndex, index int) map[string]int {
+		accum[common.Bytes2Hex(next.Validator.Validator.WithdrawalCredentials[12:])] = 1
+		return accum
+	}, map[string]int{}))
+
+	fmt.Printf("Querying %d beacon-chain withdrawal addresses to see if they may be eigenpods\n", len(interestingWithdrawalAddresses))
+
+	podOwners, err := multicall.DoManyAllowFailures[common.Address](args.Mc, lo.Map(interestingWithdrawalAddresses, func(address string, index int) *multicall.MultiCallMetaData[common.Address] {
```
Comment (gpsanant) [inline script/releases/v1.0.0-slashing/cleanup/script.go:238]: failures only include reverts, right? not rpc fails

File: script/releases/v1.0.0-slashing/cleanup/script.go [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+		success := (*podOwners)[i].Success
+		if success {
+			podToPodOwner[address] = (*podOwners)[i].Value
+		}
+		return success
+	})
+
+	fmt.Printf("Querying %d addresses on (EigenPodManager=%s) to see if it knows about these eigenpods\n", len(addressesWithPodOwners), args.PodManagerAddress)
+
+	eigenpodForOwner, err := multicall.DoManyAllowFailures(
```
Comment (gpsanant) [inline script/releases/v1.0.0-slashing/cleanup/script.go:264]: why allow failures here? nothing here should revert

File: script/releases/v1.0.0-slashing/cleanup/script.go (hunk 5 above)
Comment (jbrower95) [reply script/releases/v1.0.0-slashing/cleanup/script.go:250]: i think it's both, that's a good point though. practically for the sensitivity of testnet and mainnet we can re-run the script and validate that 0 pods have issues

File: script/releases/v1.0.0-slashing/cleanup/script.go (hunk 1 above)
Comment (jbrower95) [reply script/releases/v1.0.0-slashing/cleanup/script.go:105]: 100%

**Comment by jbrower95** [comment]
we'll review one more time hollistically as part of `jb/slashing-upgrade`. Going to merge for now.

**Pull request merged**